File: regions/system/library/dart/telemetry/test/telemetry_test.dart
import 'package:logging/logging.dart';
import 'package:test/test.dart';

import 'package:k1s0_telemetry/telemetry.dart';

void main() {
  group('TelemetryConfig', () {
    test('必須フィールドで設定が作成されること', () {
      final cfg = TelemetryConfig(
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
      );

      expect(cfg.serviceName, 'test-service');
      expect(cfg.version, '1.0.0');
      expect(cfg.tier, 'system');
      expect(cfg.environment, 'dev');
      expect(cfg.traceEndpoint, isNull);
      expect(cfg.sampleRate, 1.0);
      expect(cfg.logLevel, 'info');
    });

    test('全フィールドで設定が作成されること', () {
      final cfg = TelemetryConfig(
        serviceName: 'order-server',
        version: '2.0.0',
        tier: 'service',
        environment: 'prod',
        traceEndpoint: 'otel-collector:4317',
        sampleRate: 0.1,
        logLevel: 'warn',
      );

      expect(cfg.serviceName, 'order-server');
      expect(cfg.version, '2.0.0');
      expect(cfg.tier, 'service');
      expect(cfg.environment, 'prod');
      expect(cfg.traceEndpoint, 'otel-collector:4317');
      expect(cfg.sampleRate, 0.1);
      expect(cfg.logLevel, 'warn');
    });
  });

  group('initTelemetry', () {
    test('debugログレベルでLogger.rootがFINEに設定されること', () {
      final cfg = TelemetryConfig(
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
        logLevel: 'debug',
      );

      initTelemetry(cfg);
      expect(Logger.root.level, Level.FINE);
    });

    test('infoログレベルでLogger.rootがINFOに設定されること', () {
      final cfg = TelemetryConfig(
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'staging',
        logLevel: 'info',
      );

      initTelemetry(cfg);
      expect(Logger.root.level, Level.INFO);
    });

    test('warnログレベルでLogger.rootがWARNINGに設定されること', () {
      final cfg = TelemetryConfig(
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'prod',
        logLevel: 'warn',
      );

      initTelemetry(cfg);
      expect(Logger.root.level, Level.WARNING);
    });

    test('errorログレベルでLogger.rootがSEVEREに設定されること', () {
      final cfg = TelemetryConfig(
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'prod',
        logLevel: 'error',
      );

      initTelemetry(cfg);
      expect(Logger.root.level, Level.SEVERE);
    });

    test('未知のログレベルでINFOにデフォルト設定されること', () {
      final cfg = TelemetryConfig(
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
        logLevel: 'unknown',
      );

      initTelemetry(cfg);
      expect(Logger.root.level, Level.INFO);
    });

    test('構造化ログが出力されること', () {
      final logEntries = <LogRecord>[];
      Logger.root.onRecord.listen(logEntries.add);

      final cfg = TelemetryConfig(
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
        logLevel: 'debug',
      );

      initTelemetry(cfg);

      final logger = createLogger(cfg);
      logger.info('Test message');

      expect(logEntries, isNotEmpty);
      expect(logEntries.last.message, 'Test message');
      expect(logEntries.last.loggerName, 'test-service');
    });
  });

  group('createLogger', () {
    test('設定のサービス名でLoggerが作成されること', () {
      final cfg = TelemetryConfig(
        serviceName: 'MyService',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
      );
      final logger = createLogger(cfg);
      expect(logger.name, 'MyService');
    });

    test('異なる設定で異なるLoggerが作成されること', () {
      final cfg1 = TelemetryConfig(
        serviceName: 'ServiceA',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
      );
      final cfg2 = TelemetryConfig(
        serviceName: 'ServiceB',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
      );
      final logger1 = createLogger(cfg1);
      final logger2 = createLogger(cfg2);
      expect(logger1.name, isNot(logger2.name));
    });

    test('機能するLoggerが返されること', () {
      final cfg = TelemetryConfig(
        serviceName: 'FunctionalTest',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
      );
      final logger = createLogger(cfg);
      expect(logger, isA<Logger>());
    });
  });
}

File: regions/system/library/typescript/telemetry/__tests__/telemetry.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';

// @opentelemetry/sdk-node と @opentelemetry/exporter-trace-otlp-grpc は
// テスト環境では外部依存となるためモックする。
vi.mock('@opentelemetry/sdk-node', () => {
  const mockShutdown = vi.fn().mockResolvedValue(undefined);
  const mockStart = vi.fn();
  return {
    NodeSDK: vi.fn().mockImplementation(() => ({
      start: mockStart,
      shutdown: mockShutdown,
    })),
  };
});

vi.mock('@opentelemetry/exporter-trace-otlp-grpc', () => ({
  OTLPTraceExporter: vi.fn().mockImplementation(() => ({})),
}));

vi.mock('@opentelemetry/api', () => ({
  trace: {
    getActiveSpan: vi.fn().mockReturnValue(undefined),
  },
}));

vi.mock('pino', () => {
  const mockLogger = {
    info: vi.fn(),
    warn: vi.fn(),
    error: vi.fn(),
    debug: vi.fn(),
    level: 'info',
  };
  return {
    default: vi.fn().mockReturnValue(mockLogger),
  };
});

import { initTelemetry, shutdown, type TelemetryConfig } from '../src/telemetry';
import { createLogger } from '../src/logger';
import { httpMiddleware } from '../src/middleware';
import { NodeSDK } from '@opentelemetry/sdk-node';
import pino from 'pino';

describe('telemetry', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('initTelemetry', () => {
    it('should not create SDK when traceEndpoint is not provided', () => {
      const cfg: TelemetryConfig = {
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
        logLevel: 'debug',
      };

      initTelemetry(cfg);
      expect(NodeSDK).not.toHaveBeenCalled();
    });

    it('should create SDK when traceEndpoint is provided', () => {
      const cfg: TelemetryConfig = {
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
        traceEndpoint: 'http://localhost:4317',
        sampleRate: 1.0,
        logLevel: 'info',
      };

      initTelemetry(cfg);
      expect(NodeSDK).toHaveBeenCalled();
    });
  });

  describe('shutdown', () => {
    it('should resolve when no SDK is initialized', async () => {
      await expect(shutdown()).resolves.toBeUndefined();
    });
  });

  describe('createLogger', () => {
    it('should create a pino logger with correct config', () => {
      const cfg: TelemetryConfig = {
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
        logLevel: 'debug',
      };

      const logger = createLogger(cfg);
      expect(pino).toHaveBeenCalledWith(
        expect.objectContaining({
          level: 'debug',
          base: {
            service: 'test-service',
            version: '1.0.0',
            tier: 'system',
            environment: 'dev',
          },
          mixin: expect.any(Function),
        }),
      );
      expect(logger).toBeDefined();
    });

    it('should set info log level for staging', () => {
      const cfg: TelemetryConfig = {
        serviceName: 'staging-service',
        version: '2.0.0',
        tier: 'business',
        environment: 'staging',
        logLevel: 'info',
      };

      createLogger(cfg);
      expect(pino).toHaveBeenCalledWith(
        expect.objectContaining({
          level: 'info',
          base: expect.objectContaining({
            service: 'staging-service',
            environment: 'staging',
          }),
        }),
      );
    });

    it('should set warn log level for prod', () => {
      const cfg: TelemetryConfig = {
        serviceName: 'prod-service',
        version: '3.0.0',
        tier: 'service',
        environment: 'prod',
        logLevel: 'warn',
      };

      createLogger(cfg);
      expect(pino).toHaveBeenCalledWith(
        expect.objectContaining({
          level: 'warn',
          base: expect.objectContaining({
            service: 'prod-service',
            environment: 'prod',
          }),
        }),
      );
    });
  });

  describe('httpMiddleware', () => {
    it('should return a middleware function', () => {
      const cfg: TelemetryConfig = {
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
        logLevel: 'debug',
      };

      const logger = createLogger(cfg);
      const middleware = httpMiddleware(logger);
      expect(typeof middleware).toBe('function');
    });

    it('should call next function', () => {
      const cfg: TelemetryConfig = {
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
        logLevel: 'debug',
      };

      const logger = createLogger(cfg);
      const middleware = httpMiddleware(logger);
      const next = vi.fn();
      const req = { method: 'GET', url: '/health' } as any;
      const res = {
        statusCode: 200,
        on: vi.fn(),
      } as any;

      middleware(req, res, next);
      expect(next).toHaveBeenCalled();
    });

    it('should log on response finish', () => {
      const cfg: TelemetryConfig = {
        serviceName: 'test-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
        logLevel: 'debug',
      };

      const logger = createLogger(cfg);
      const middleware = httpMiddleware(logger);
      const next = vi.fn();
      const req = { method: 'POST', url: '/api/v1/orders' } as any;
      let finishCallback: (() => void) | undefined;
      const res = {
        statusCode: 201,
        on: vi.fn((event: string, cb: () => void) => {
          if (event === 'finish') {
            finishCallback = cb;
          }
        }),
      } as any;

      middleware(req, res, next);
      expect(finishCallback).toBeDefined();

      finishCallback!();
      expect(logger.info).toHaveBeenCalledWith(
        expect.objectContaining({
          method: 'POST',
          path: '/api/v1/orders',
          status: 201,
        }),
        'Request completed',
      );
    });
  });

  describe('TelemetryConfig', () => {
    it('should accept all configuration fields', () => {
      const cfg: TelemetryConfig = {
        serviceName: 'my-service',
        version: '1.0.0',
        tier: 'business',
        environment: 'staging',
        traceEndpoint: 'http://otel-collector:4317',
        sampleRate: 0.5,
        logLevel: 'info',
      };

      expect(cfg.serviceName).toBe('my-service');
      expect(cfg.version).toBe('1.0.0');
      expect(cfg.tier).toBe('business');
      expect(cfg.environment).toBe('staging');
      expect(cfg.traceEndpoint).toBe('http://otel-collector:4317');
      expect(cfg.sampleRate).toBe(0.5);
      expect(cfg.logLevel).toBe('info');
    });

    it('should allow optional traceEndpoint and sampleRate', () => {
      const cfg: TelemetryConfig = {
        serviceName: 'simple-service',
        version: '1.0.0',
        tier: 'system',
        environment: 'dev',
        logLevel: 'debug',
      };

      expect(cfg.traceEndpoint).toBeUndefined();
      expect(cfg.sampleRate).toBeUndefined();
    });
  });
});
